### PR TITLE
update weights from distributed

### PR DESCRIPTION
**Pull request by zhaochenyang20**
## Motivation

Following [OpenRLHF's vllm wrapper](https://github.com/OpenRLHF/OpenRLHF/blob/main/openrlhf/trainer/ray/vllm_worker_wrap.py), we support online weights updates from `torch.distributed`, which is hundreds of times quicker than saving the model to disk and updating the engine from disk.

Following these PRs:

https://github.com/sgl-project/sglang/pull/2265

https://github.com/sgl-project/sglang/pull/2266

## Modifications

I added three APIs in the server and engine, and added tests for them:

1. `init_parameter_update_group`
2. `update_parameter_from_distributed`
3. `get_weights_by_parameter_name`

## Checklist

- [x] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [x] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [x] Update documentation as needed, including docstrings or example tutorials.

### CODE REVIEW

File: python/sglang/srt/managers/scheduler.py [hunk 1]
```diff
@@ -33,18 +33,25 @@
 from sglang.srt.hf_transformers_utils import get_processor, get_tokenizer
 from sglang.srt.layers.logits_processor import LogitsProcessorOutput
 from sglang.srt.managers.io_struct import (
+    HEAD,
```
Comment (merrymercy) [inline python/sglang/srt/managers/scheduler.py]: remove this

File: python/sglang/srt/managers/tokenizer_manager.py [hunk 2]
```diff
@@ -37,6 +38,7 @@
     get_image_processor,
 )
 from sglang.srt.managers.io_struct import (
+    HEAD,
```
Comment (merrymercy) [inline python/sglang/srt/managers/tokenizer_manager.py]: what is this?

File: python/sglang/srt/managers/tokenizer_manager.py [hunk 3]
```diff
@@ -451,6 +461,80 @@ async def update_weights(
...
+        if self.to_create_loop:
+            self.create_handle_loop()
+
+        if obj.backend is None:
+            obj.backend = "nccl"
+        self.send_to_scheduler.send_pyobj(obj)
+
+        self.init_parameter_update_group_result = asyncio.Future()
+
+        if self.server_args.dp_size == 1:
```
Comment (merrymercy) [inline python/sglang/srt/managers/tokenizer_manager.py]: it seems this only support dp_size = 1, so you can `assert self.server_args.dp_size == 1` here

File: python/sglang/srt/model_executor/model_runner.py [hunk 4]
```diff
@@ -400,6 +404,106 @@ def model_load_weights(model, iter):
...
+
+        rank = rank_offset + self.tp_rank
+
+        logger.info(
+            f"init custom process group: master_address={master_address}, master_port={master_port}, "
+            f"rank_offset={rank_offset}, world_size={world_size}, group_name={group_name}, backend={backend}"
+        )
+
+        try:
+            self._model_update_group = init_custom_process_group(
```
Comment (merrymercy) [inline python/sglang/srt/model_executor/model_runner.py:433]: do not use `init_custom_process_group`. Use `torch.dist.init`

File: python/sglang/srt/server.py [hunk 5]
```diff
@@ -210,6 +214,54 @@ async def update_weights(obj: UpdateWeightReqInput, request: Request):
...
+    )
+    content = {"success": success, "message": message}
+    if success:
+        return ORJSONResponse(content, status_code=200)
+    else:
+        return ORJSONResponse(content, status_code=HTTPStatus.BAD_REQUEST)
+
+
+@app.api_route("/get_weights_by_parameter_name", methods=["GET", "POST"])
+async def get_weights_by_parameter_name(
```
Comment (merrymercy) [inline python/sglang/srt/server.py]: rename:
- `init_parameter_update_group` -> `init_weight_update_group`
- `update_parameter_from_distributed ` -> `update_weights_from_distributed`
- `get_weights_by_parameter_name` -> `get_weights_by_name`

**Review by merrymercy — CHANGES_REQUESTED** [review]
